### PR TITLE
Apply code style rules to soss_extract

### PR DESCRIPTION
**Pull request by emolter**
Partially resolves [JP-3860](https://jira.stsci.edu/browse/JP-3860)

This PR updates the `soss_extract` submodule to comply with the new code style rules.

Since `soss_extract` is nestled inside `extract_1d`, I had to add some ugly per-file-ignores to `.pre-commit-config.yaml` and `.ruff.toml`; these can be removed once the rest of `extract_1d` is updated to follow the style rules.

As a procedural note, I added a line to the epic labeled `soss_extract` and assigned it to myself, then removed my name from the assignment list next to `extract_1d` so someone else could pick up the rest of the step.

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

File: jwst/extract_1d/soss_extract/atoca_utils.py [hunk 1]
```diff
@@ -798,13 +823,16 @@ def ThroughputSOSS(wavelength, throughput):
...
+
     def interpolator(wv):
         wv = np.clip(wv, wl_min, wl_max)
         return interp(wv)
+
     return interpolator
 
 
 class WebbKernel:
+    """The JWST kernel."""
```
Comment (melanieclarke) [inline jwst/extract_1d/soss_extract/atoca_utils.py:835]: A little more explanation would be nice here, if we can. Is this a PSF kernel?

File: jwst/extract_1d/soss_extract/atoca_utils.py [hunk 2]
```diff
@@ -967,7 +999,8 @@ def __call__(self, wave, wave_c):
 
 
 def _constant_kernel_to_2d(c, grid_range):
-    """Build a 2d kernel array with a constant 1D kernel as input
+    """
+    Build a 2d kernel array with a constant 1D kernel as input.
```
Comment (melanieclarke) [inline jwst/extract_1d/soss_extract/atoca_utils.py:1003]:
```suggestion
    Build a 2D kernel array with a constant 1D kernel as input.
```

File: jwst/extract_1d/soss_extract/atoca_utils.py [hunk 3]
```diff
@@ -1622,94 +1657,94 @@ def __init__(self, test_dict, default_chi2='chi2_cauchy'):
...
-        # retrieve loss function
         try:
             loss = LOSS_FUNCTIONS[loss]
-        except KeyError as e:
-            msg = (f"loss={loss} not a valid key."
-                   f"Must be one of {[LOSS_FUNCTIONS.keys()]} or callable.")
-            raise e(msg)
+        except KeyError:
+            msg = (
+                f"loss={loss} not a valid key.Must be one of {[LOSS_FUNCTIONS.keys()]} or callable."
```
Comment (melanieclarke) [inline jwst/extract_1d/soss_extract/atoca_utils.py:1686]:
```suggestion
                f"loss={loss} not a valid key. Must be one of {[LOSS_FUNCTIONS.keys()]} or callable."
```

**Review by melanieclarke — COMMENTED** [review]
Lots here, thanks for getting it cleaned up.  I can take the rest of the extract1d module, when this one is in.

I spotted a few typos and other small things while reading through.

File: jwst/extract_1d/soss_extract/atoca.py [hunk 4]
```diff
@@ -128,8 +143,10 @@ def __init__(self, wave_map, trace_profile, throughput, kernels,
         self.orders = orders
         self.n_orders = len(self.orders)
         if self.n_orders != len(self.wave_map):
-            msg = ("The number of orders specified ({}) and the number of "
-                   "wavelength maps provided ({}) do not match.")
+            msg = (
+                "The number of orders specified ({}) and the number of "
+                "wavelength maps provided ({}) do not match."
```
[the remark this answers is not in the document]
Comment (emolter) [reply jwst/extract_1d/soss_extract/atoca.py:148]: I have no idea, I guess someone started typing this f-string and then forgot to finish?  Thanks for catching it

File: jwst/extract_1d/soss_extract/atoca_utils.py (hunk 1 above)
Comment (emolter) [reply jwst/extract_1d/soss_extract/atoca_utils.py]: I tried my best, but to be honest someone from the INS team should probably write this.  I know @hover2pi  has general documentation improvements for ATOCA on his radar.  I also know he has his Slack status set to illness

Comment (melanieclarke) [reply jwst/extract_1d/soss_extract/atoca_utils.py]: Okay, let's defer this then.

**Review by melanieclarke — APPROVED** [review]
I think these changes are uncontroversial enough we can go ahead and get them merged.  We'll follow up on more documentation updates later.
